arangod/Aql/JoinExecutor.cpp

```diff
@@ -102,9 +83,55 @@ auto JoinExecutor::produceRows(AqlItemBlockInputRange& inputRange,
           _projectionsBuilder.close();
           output.moveValueInto(_infos.indexes[k].documentOutputRegister,
                                _currentRow, _projectionsBuilder.slice());
-        }
 
-        projectionsOffset += idx.projections.size();
+          projectionsOffset += idx.projections.size();
+        } else {
+          // load the whole document
+
+          // TODO post filter based on document value
+          // TODO do we really want to check/populate cache?
+          _infos.indexes[k].collection->getCollection()->getPhysical()->lookup(
+              &_trx, docIds[k],
+              {[&](LocalDocumentId token, VPackSlice doc) {
+                 if (idx.projections.empty()) {
+                   output.moveValueInto(
+                       _infos.indexes[k].documentOutputRegister, _currentRow,
+                       doc);
+                 } else {
+                   _projectionsBuilder.clear();
+                   _projectionsBuilder.openObject(true);
+                   idx.projections.toVelocyPackFromDocument(_projectionsBuilder,
+                                                            doc, &_trx);
+                   _projectionsBuilder.close();
+                   output.moveValueInto(
+                       _infos.indexes[k].documentOutputRegister, _currentRow,
+                       _projectionsBuilder.slice());
+                 }
+
+                 return true;
+               },
+               [&](LocalDocumentId token, std::unique_ptr<std::string>& doc) {
+                 if (idx.projections.empty()) {
+                   output.moveValueInto(
+                       _infos.indexes[k].documentOutputRegister, _currentRow,
+                       &doc);
+                 } else {
+                   _projectionsBuilder.clear();
+                   _projectionsBuilder.openObject(true);
+                   idx.projections.toVelocyPackFromDocument(
+                       _projectionsBuilder,
+                       VPackSlice(
+                           reinterpret_cast<uint8_t const*>(doc->data())),
+                       &_trx);
+                   _projectionsBuilder.close();
+                   output.moveValueInto(
+                       _infos.indexes[k].documentOutputRegister, _currentRow,
+                       _projectionsBuilder.slice());
```

Comment:
Maybe better to doesn't make projection for such case and keep whole document?
Btw velocypack::Buffer can be moved to aqlvalue, so you can use bufferRef for this

Reply:
You are certainly correct for single server use cases, but as soon as we send that document over the network, you want to have the projected value instead.